### ISSUE TITLE
Understand debugging use cases

### ISSUE THREAD

**Issue by yetudada** [kedro-org/kedro-viz#1519]
## Introduction
Some of our users have indicated that they use Kedro-Viz for debugging and want us to extend functionality there, here are some quotes to validate this from @NeroOkwa in #1464: 

“And in case something fails, it will be amazing if you can say I know that in this, at this point, something fails, so this box, it's red [indicates node on Viz]”.
“I'm always debugging because I'm always running the whole pipeline. When you see the error in Kedro, it's not always clear where something is failing. So if you can validate that quickly in Kedro-Viz, it'll be super useful".

“I think there's something around debugging… something like this [code preview] is probably more helpful than if we're just gonna go line by line through the code so that I can understand what part of the code are we running into issues and then help".

Using Kedro-Viz to debug a Kedro project - “I think an unintended use, for example, when my team tell me the project is ready, and the first thing I do is to plot the repository, install the requirement, open Kedro-Viz. If there's an error and I can run it, I know that there's a problem in the catalog in the pipe somewhere, but that's one of things I use Kedro-Viz for".

“So there were times when Kedro-Viz, obviously it needs all of the dependencies to be working correctly in order to generate the, the visualisation. At times, it felt like it would be valuable if those dependencies weren't accurate. If those pieces could be orphaned as almost like troubleshooting where the pipeline might be going wrong. It's more nice to have, obviously the pipeline should just be right, but it felt like it might be a cool opportunity".
“This would help to track down upstream and downstream the missing dependencies or where something might be broken in a pipeline to make troubleshooting easier using DAGs instead of the IDEs ".

Copying the initial user comment and feature request on the [slack channel](https://kedro-org.slack.com/archives/C03RKP2LW64/p1688630506702459) here:
"I want to log the number of rows for the datasets at each step of my pipeline. It's for debugging. The goal is to notice big drop of rows during one data transformation step. For example, after one node, I may see that my number of lines drops by 30% when it’s supposed to stay the same."

## What needs to be done? 
This is first a research task to understand: 
 - What do users mean by debugging? 
 - What type of errors do they typically run into when their pipeline fails? 
 - Why can't they tell why things have failed via logs on the CLI? And how do they figure out that something is wrong? _This one is important because if no part of Kedro e.g. logs, can tell you what's wrong then there's no way that Kedro-Viz will know too_

> If possible, ask users to show you how they debug 🙏🏾 

We need to know the categories of errors users run into because we can then work out what's the best way to help them e.g. if we know that 62% of users experience pipeline errors because they have a bad table join then how could we best display that information? Or if 31% of users experience pipeline failure because an upstream dataset is missing, then what would be the best way to let them know that on Kedro-Viz?

**Comment by astrojuanlu** [comment]
An update on this, I hadn't make it public.

We kicked off this research on October 14th by looking into our `#questions` channel in Slack (archived in https://kedro.hall.community/ and https://linen-slack.kedro.org/) as well as GitHub issues opened by the community.

Our initial hypothesis was that this would help us understand the kind of issues highlighted by @NeroOkwa and @yetudada, for example (from https://github.com/kedro-org/kedro-viz/issues/1519#issue-1877159148):

> “And in case something fails, it will be amazing if you can say I know that in this, at this point, something fails, so this box, it's red [indicates node on Viz]”.
> “I'm always debugging because I'm always running the whole pipeline. When you see the error in Kedro, it's not always clear where something is failing. So if you can validate that quickly in Kedro-Viz, it'll be super useful".
> “I think there's something around debugging… something like this [code preview] is probably more helpful than if we're just gonna go line by line through the code so that I can understand what part of the code are we running into issues and then help".
> “I think an unintended use, for example, when my team tell me the project is ready, and the first thing I do is to plot the repository, install the requirement, open Kedro-Viz. If there's an error and I can run it, I know that there's a problem in the catalog in the pipe somewhere, but that's one of things I use Kedro-Viz for".
> “So there were times when Kedro-Viz, obviously it needs all of the dependencies to be working correctly in order to generate the, the visualisation. At times, it felt like it would be valuable if those dependencies weren't accurate. If those pieces could be orphaned as almost like troubleshooting where the pipeline might be going wrong. It's more nice to have, obviously the pipeline should just be right, but it felt like it might be a cool opportunity".
> “This would help to track down upstream and downstream the missing dependencies or where something might be broken in a pipeline to make troubleshooting easier using DAGs instead of the IDEs ".

However, we found that most of the issues reported by users had to do with Kedro:

```
┌──────────────────────────────┬───────┐
│ Tags                         ┆ count │
│ ---                          ┆ ---   │
│ str                          ┆ u32   │
╞══════════════════════════════╪═══════╡
│ tool: tensorflow             ┆ 2     │
│ plugin: kedro-datasets       ┆ 2     │
│ component: logging           ┆ 2     │
│ component: runners           ┆ 3     │
│ plugin: kedro-viz            ┆ 3     │
│ component: dataset factories ┆ 3     │
│ cloud: aws                   ┆ 3     │
│ component: credentials       ┆ 5     │
│ component: config loader     ┆ 10    │
│ component: i/o               ┆ 10    │
└──────────────────────────────┴───────┘
```

This made us realise that we need to dig a bit deeper to learn how people debug their own code and how Kedro and Kedro Viz can help.

And yet, looking at prior external research, we found that

- There is a general "lack of knowledge regarding developers’ debugging behavior"
- Despite the existence of sophisticated debugging tools, most people use "print-style" debugging
- Inquiring into people's debugging workflows requires "letting others do research on how one approaches what are essentially one’s own faults is a precarious situation for both a developer and a researcher"

(Insights coming from Beller, M. *et al.* (2018) ‘On the dichotomy of debugging behavior among programmers’, *Proceedings of the 40th International Conference on Software Engineering* [Preprint]. doi:10.1145/3180155.3180175)

And, additionally,

- There is little research on how people debug their _data_ ("When a program fails to process an input, it need not be the program code that is at fault. It can also be that the input data is faulty, for instance as result of data corruption.")

(Kirschner, L., Soremekun, E. and Zeller, A. (2020) ‘Debugging inputs’, *Proceedings of the ACM/IEEE 42nd International Conference on Software Engineering*, 1, pp. 75–86. doi:10.1145/3377811.3380329.)

In conclusion, we decided to

- [x] Extend the current workflow of logging user support requests to the Wizard responsibilities (done ✔️ see https://github.com/orgs/kedro-org/projects/10 and the different automations @merelcht has implemented around it)
- [ ] Turn this into an "etnographic research" happening some time in 2025

**Comment by astrojuanlu** [comment]
Part of this is being tackled as part of https://github.com/kedro-org/kedro-viz/issues/2310 Closing this issue in favour of that one.
